### DISCUSSION TITLE
Can gym-ignition export SDF?

### DISCUSSION THREAD

**Discussion by FirefoxMetzger** [robotology/gym-ignition#394] [Q&A]
Gym-ignition offers various ways to insert models and worlds from SDF (which is awesome!). I was wondering if the opposite is possible. Assuming I have a simulator in some state, can I extract a SDF that represents the current world?

I know this can be done manually in the Ignition GUI, but I am particularly interested in doing it programmatically.

My use-case is that I want to use the SDF to synchronize a pose-graph structure that I wrote on top of NumPy with the current simulator's state. The motivation for this is that I am writing a custom IK solution for my research, and it is easier/more hackable in python. I will then feed the output back into the simulator as a target for the controller, but for the entire loop to work, I need access to the current simulator state.

Alternatively, I could manually synchronize available joints and model positions explicitly; however, that approach seems more complex and less general than simply extracting the SDF (or rather the `//state` element of the SDF to be precise), in particular in the context of gym-ignition inserting or removing models at runtime (which I would like to be aware of).

**Comment by diegoferigo** [comment]
At the moment no, this feature is not possible with ScenarIO APIs. However, as you wrote, it seems that there's somewhere in upstream the support to export worlds. I found the [`SdfGenerator.hh`](https://github.com/ignitionrobotics/ign-gazebo/blob/main/src/SdfGenerator.hh) file in the codebase, however it is not public. If you're interested and you manage to find a way to use these methods, feel free to propose a PR, it could be a great addition!

Something to keep in mind is checking whether the models are serialized correctly in the case their properties are changed dynamically from Python (e.g. the joint velocity / torque limits to make an example). If not, we can make it clear from the documentation of the new functionality.
